x/virtualgroup/keeper/keeper.go

```diff
@@ -0,0 +1,684 @@
+	store := ctx.KVStore(k.storeKey)
+
+	bz := k.cdc.MustMarshal(gvgsStatisticsWithinSP)
+
+	store.Set(types.GetGVGStatisticsWithinSPKey(gvgsStatisticsWithinSP.StorageProviderId), bz)
+}
+
+func (k Keeper) BatchSetGVGStatisticsWithinSP(ctx sdk.Context, gvgsStatisticsWithinSP []*types.GVGStatisticsWithinSP) {
+	for _, g := range gvgsStatisticsWithinSP {
+		k.SetGVGStatisticsWithSP(ctx, g)
+	}
+}
+
+func (k Keeper) IsStorageProviderCanExit(ctx sdk.Context, spID uint32) error {
+	store := ctx.KVStore(k.storeKey)
+
+	prefixStore := prefix.NewStore(store, types.GetGVGFamilyPrefixKey(spID))
+	iter := prefixStore.Iterator(nil, nil)
+	if iter.Valid() {
+		var family types.GlobalVirtualGroupFamily
+		k.cdc.MustUnmarshal(iter.Value(), &family)
+		return types.ErrSPCanNotExit.Wrapf("not swap out from all the family, key: %s", family.String())
+	}
+
+	gvgStat, found := k.GetGVGStatisticsWithinSP(ctx, spID)
+	if found && gvgStat.SecondaryCount != 0 {
+		return types.ErrSPCanNotExit.Wrapf("not swap out from all the gvgs, remain: %d", gvgStat.SecondaryCount)
+	}
+	return nil
+}
+
+func (k Keeper) RebindingGVGsToBucket(ctx sdk.Context, bucketID math.Uint, dstSP *sptypes.StorageProvider, newLVGToGVGMappings []*storagetypes.LVGToGVGMapping) error {
+	gvgsBindingOnBucket, found := k.GetGVGsBindingOnBucket(ctx, bucketID)
+	if !found {
+		// empty bucket. do nothing
+		return nil
+	}
+	var newGVGBindingOnBucket types.GlobalVirtualGroupsBindingOnBucket
+	lvg2gvg := make(map[uint32]uint32)
+	for i, lvgID := range gvgsBindingOnBucket.LocalVirtualGroupIds {
+		lvg2gvg[lvgID] = gvgsBindingOnBucket.GlobalVirtualGroupIds[i]
+	}
+
+	// verify secondary signature
+	for _, newLvg2gvg := range newLVGToGVGMappings {
+		dstGVG, found := k.GetGVG(ctx, newLvg2gvg.GlobalVirtualGroupId)
+		if !found {
+			return types.ErrGVGNotExist.Wrapf("dst gvg not found")
+		}
+
+		lvg, found := k.GetLVG(ctx, bucketID, newLvg2gvg.LocalVirtualGroupId)
+		if !found {
+			return types.ErrGVGNotExist.Wrapf("lvg not found")
+		}
+
+		srcGVGID, exists := lvg2gvg[newLvg2gvg.LocalVirtualGroupId]
+		if !exists {
+			return types.ErrMigrationBucketFailed.Wrapf("local virtual group not found, id: %d", newLvg2gvg.LocalVirtualGroupId)
+		}
+
+		srcGVG, found := k.GetGVG(ctx, srcGVGID)
+		if !found {
+			return types.ErrGVGNotExist.Wrapf("src gvg not found, ID: %d", srcGVGID)
+		}
+
+		for i, sspID := range dstGVG.SecondarySpIds {
+			ssp, found := k.spKeeper.GetStorageProvider(ctx, sspID)
+			if !found {
+				return sptypes.ErrStorageProviderNotFound.Wrapf("spID: %d", sspID)
+			}
+
+			migrationBucketSignDoc := types.NewMigrationBucketSignDoc(
+				bucketID, dstSP.Id, newLvg2gvg.LocalVirtualGroupId, srcGVGID, newLvg2gvg.GlobalVirtualGroupId)
+			err := types2.VerifySignature(sdk.MustAccAddressFromHex(ssp.ApprovalAddress), sdk.Keccak256(migrationBucketSignDoc.GetSignBytes()), newLvg2gvg.SecondarySpSignature[i])
```

Comment:
Why not use bls signature here?

Reply:
We use bls aggregate signatures for multiple secondary SPs within GVG, but not between GVGs, Because the signature content is different and secondary sp does not need to know about the migration of other GVGS